impl/src/main/java/org/jboss/weld/context/beanstore/AttributeBeanStore.java

```diff
@@ -80,7 +84,7 @@ public boolean detach() {
     public boolean attach() {
         if (!attached) {
             attached = true;
-            if (isLocalBeanStoreSyncNeeded()) {
+            if (isLocalBeanStoreSyncNeeded() && !isAttributeLazyLoadingAllowed()) {
                 // The local bean store is authoritative, so copy everything to the backing store
                 for (BeanIdentifier id : beanStore) {
```

Comment:
I think that this logic, unlike the readAttributes() logic, should be executed regardless lazy loading is allowed or not. In other words, we are loading instances from the backing store to the local bean store lazily but we are writing all the local entries to the backing store on attach(). This should not impact performance much as the local store will be empty most of the time. Also, it should make reasoning about the correctness of this easier.